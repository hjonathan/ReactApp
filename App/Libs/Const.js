import RNFS from "react-native-fs";

export const CASE = {
    INBOX: "Inbox",
    DRAFT: "Draft",
    PARTICIPATED: "Participated",
    UNASSIGNED: "Unassigned",
    STATUS: {
        NONE: "none",
        SENDING: "sending",
        WORKING: "working",
        OFFLINE: "offline"
    }
};

export const DATETIME = {
    LOCAL: "YYYY-MM-DD HH:mm:ss",
    LOCAL_UTC: "YYYY-MM-DD HH:mm:ss Z"
};

export const DATE = {
    SIMPLE: "YYYYMMDD",
    LOCAL: "YYYY-MM-DD"
};

export const TIME = {
    LOCAL: "HH:mm:ss"
};

export const DS = "/";

export const PATH = {
    PROJECT: `${RNFS.DocumentDirectoryPath}`,
    FORMRENDER: `${DS}FormRender`,
    STEPS: `${DS}Steps`,
    NEXTSTEPS: `${DS}NextSteps`,
    FROMS: `${DS}Forms`,
    DATA: `${DS}Data`,
    EXTERNALLIBS: `${DS}ExternalLibs`,
    BUILDPROD: `${DS}build-prod`,
    MUSIC: "/storage/emulated/0/Music/processmakeraudiorecord/"
};

export const FILENAMES = {
    EXTERNALMAP: "ExtMap.json"
};
