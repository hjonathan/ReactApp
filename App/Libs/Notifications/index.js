/* eslint-disable import/prefer-default-export */
import * as pushNotification from "./pushNotification";

export {pushNotification};
