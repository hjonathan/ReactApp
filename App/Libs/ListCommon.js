import _ from "lodash";
import Task from "../Model/Task";
import Step from "../Model/Step";
import {CASE} from "./Const";

/**
 * Function commons for: inbox, draft, unassigned and participated
 */
export default class ListCommon {
    /**
     * Constructor
     * @param props
     */
    constructor (props) {
        this.props = props;
        this.task = Task.allTaskIdOffline();
    }

    /**
     * Get data lists
     * @param method
     * @param params
     */
    getDataList (method, params = {}) {
        if (this.props[method]) {
            this.props[method](params);
        }
    }

    /**
     * Gets usersInfo
     * @param caseList
     * @param type is type in list: prevUser, currentUser
     */
    getUsersInfo (caseList, type) {
        if (caseList.length > 0 && caseList[0][type]) {
            const ids = caseList.map(cases => cases[type].userId);
            this.props.userInfoRequest({
                size: "1",
                user: {
                    ids: _.uniq(ids)
                }
            });
        }
    }

    /**
     * Merges userPhoto props to dataList
     * @param params
     * @param type is type in list: prevUser, currentUser
     */
    mergeList (params, type) {
        let i,
            j,
            cbUser = function (user) { return user.userId === params.caseList[i][type].userId; };
        if (params.caseList) {
            for (i = 0; i < params.caseList.length; i += 1) {
                if (params.caseList[i][type]) {
                    j = _.findIndex(params.users, cbUser);
                    if (j >= 0) {
                        params.caseList[i].userPhoto = params.users[j].userPhoto;
                    }
                    params.caseList[i].status = this.processIfOffline(params.caseList[i]);
                }
            }
        }
        return params.caseList;
    }

    /**
     * Check if case is offline and get all necessary for case
     * @param caseData
     * @returns {boolean}
     */
    processIfOffline (caseData) {
        let isOffline,
            {status} = caseData;
        if (!status || status === CASE.STATUS.NONE) {
            if (caseData.task.offlineEnabled === "TRUE" && this.props.caseOffline) {
                this.props.caseOffline(caseData);
                status = CASE.STATUS.OFFLINE;
            }
            isOffline = Step.existSteps(caseData.process.processId, caseData.task.taskId);
            if (isOffline && _.includes(this.task, caseData.task.taskId)) {
                status = CASE.STATUS.OFFLINE;
            }
        }
        return status;
    }

    /**
     * Key for FlatList formed by delIndex and caseId
     * @param item
     * @returns {string}
     */
    keyExtractor (item, index) {
        return `${item.delIndex}_${item.caseId}`;
    }
}
