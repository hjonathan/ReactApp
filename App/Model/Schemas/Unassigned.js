import * as List from "./List";
import {CASE} from "../../Libs/Const";

/**
 * Schema for model Unassigned
 */
export default class UnassignedSchema {
    static schema = {
        name: "UnassignedSchema",
        primaryKey: "caseId",
        properties: {
            caseId: {type: "string"},
            delIndex: {type: "string"},
            task: {type: List.TaskSchema.schema.name},
            process: {type: List.ProcessSchema.schema.name},
            caseNumber: {type: "string"},
            caseTitle: {type: "string"},
            date: {type: "string?"},
            delegateDate: {type: "string?"},
            prevUser: {type: List.UsersSchema.schema.name},
            dueDate: {type: "string"},
            status: {type: "string?", default: CASE.STATUS.NONE}
        }
    };
}
