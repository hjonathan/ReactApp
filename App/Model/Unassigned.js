import _ from "lodash";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, UnassignedSchema} from "./Schemas";

/**
 * The methods for the model UnassignedCase
 */
class UnassignedCase extends BaseModel {
    constructor () {
        super(UnassignedSchema.schema.name);
    }
    /**
     * Return all rows Unassigned
     */
    getUnassignedOffline () {
        let result = RealmObject.objects(UnassignedSchema.schema.name);
        return result.map(x => _.assign({}, x));
    }
    /**
     * Save the list when the action cases/unassigned/success is executed
     * @param listUnassigned
     */
    saveList (listUnassigned) {
        RealmObject.write(() => {
            listUnassigned.data.map((unassignedCase) => {
                RealmObject.create(UnassignedSchema.schema.name, unassignedCase, true);
                return;
            });
        });
    }
    /**
     * Delete multiple Cases
     */
    destroyObject () {
        RealmObject.write(() => {
            let allObjects = RealmObject.objects(UnassignedSchema.schema.name);
            RealmObject.delete(allObjects); // Deletes all cases in Unassigned
        });
    }
}

export default new UnassignedCase();
