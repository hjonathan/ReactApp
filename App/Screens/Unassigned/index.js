import {connect} from "react-redux";
import Unassigned from "./unassigned";
import actions from "../../Actions";
import {CASE} from "../../Libs/Const";

const mapStateToProps = state => ({
        cases: {
            unassigned: state.Cases.unassigned || null,
            users: state.Cases.users || []
        },
        list: CASE.UNASSIGNED,
        disableItemList: state.Screens.All.disableItemList,
        isConnected: state.Net.isConnected
    }),
    mapDispatchToProps = dispatch => ({
        requestCasesUnassigned: params => dispatch(actions.cases.unassigned.request(params)),
        userInfoRequest: params => dispatch(actions.cases.userInfo.request(params)),
        routeName: name => dispatch(actions.cases.routeName(name)),
        reset: () => dispatch(actions.cases.reset())
    });

export default connect(mapStateToProps, mapDispatchToProps)(Unassigned);
