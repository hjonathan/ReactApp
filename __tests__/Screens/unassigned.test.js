import React from 'react'
import {shallow} from 'enzyme';
import Unassigned from '../../App/Screens/Unassigned';
import UnassignedScreen from '../../App/Screens/Unassigned/unassigned';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import {createStore} from 'redux';
import {CASE} from "../../App/Libs/Const";

jest.mock('react-native-orientation', () => require.requireActual('../../__mocks__/react-native-orientation').default);

describe('Screen >>> Unassigned', () => {
    const initialState = {
        Cases: {
            unassigned: [
                {
                    prevUser: {
                        userId: "00001"
                    }
                },
                {
                    prevUser: {
                        userId: "00002"
                    }
                }
            ],
            users: [
                {
                    userId: "00001",
                    userPhoto: "picture1"
                },
                {
                    userId: "00002",
                    userPhoto: "picture2"
                }
            ]
        },
        Screens: {
            All: {
                disabledItemList: false
            }
        },
        Net: {
            isConnected: true
        }
    },
        requestCasesUnassigned = jest.fn(),
        userInfoRequest = jest.fn(),
        routeName = jest.fn(),
        navigation = {
            dispatch: jest.fn(),
            setParams: jest.fn(),
            state: {
                routeName: CASE.DRAFT,
                params:{
                    searchInput: null
                }
            }
        },
        mockStore = configureStore();

    let store,
        container,
        viewUnassigned,
        viewUnassignedOffline;

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(<Unassigned store={store}/>);
        viewUnassigned = shallow(
            <UnassignedScreen
                navigation={navigation}
                isConnected={true}
                requestCasesUnassigned={requestCasesUnassigned}
                userInfoRequest={userInfoRequest}
                routeName={routeName}
                reset={jest.fn()}
                store={store}
            />);
        viewUnassignedOffline = shallow(<UnassignedScreen navigation={navigation} isConnected={false} routeName={routeName} />);
    });

    /**
     * Render test
     */
    it('Render unassigned correctly', () => {
        expect(toJson(container)).toMatchSnapshot();
    });

    it('Check static navigationOptions', () => {
        expect(UnassignedScreen.navigationOptions({navigation}).title)
            .toEqual("menu_drawer_main_menu_item_unassigned");
        expect(UnassignedScreen.navigationOptions({navigation}).headerTintColor)
            .toEqual("#FFF");
        expect(UnassignedScreen.navigationOptions({navigation}).headerStyle)
            .toEqual({backgroundColor: "#1F6CBC"});
        expect(UnassignedScreen.navigationOptions({navigation}).headerLeft).toMatchSnapshot();
    });

    it('mapStateToProps', () => {
        expect(container.props('children').requestCasesUnassigned()).toMatchSnapshot();
        expect(container.props('children').userInfoRequest()).toMatchSnapshot();
        expect(container.props('children').routeName()).toMatchSnapshot();
        expect(container.props('children').reset()).toMatchSnapshot();
    });

    it("Should call navigationOptions method", () => {
        expect(viewUnassigned.instance().navigationOptions).toBeUndefined();
    });

    it("Props coverage cases", () => {
        expect(container.prop("cases")).toEqual(initialState.Cases);
    });

    it("Props coverage disableItemList", () => {
        expect(container.prop("disableItemList")).toEqual(initialState.Screens.All.disableItemList);
    });

    it("Props coverage componentWillReceiveProps", () => {
        viewUnassigned.instance().componentWillReceiveProps({
            cases: {
                unassigned: initialState.Cases.unassigned
            },
            list: CASE.UNASSIGNED,
            disableItemList: initialState.Screens.All.disableItemList
        });
        expect(viewUnassigned.instance().state.unassignedRequested).toEqual(true);
        viewUnassigned.instance().componentWillReceiveProps({
            cases: {
                users: initialState.Cases.users
            },
            list: CASE.UNASSIGNED,
            disableItemList: initialState.Screens.All.disableItemList
        });
        viewUnassigned.instance().setState({usersRequested: false});
        viewUnassigned.instance().componentWillReceiveProps({
            cases: {
                unassigned: {response: []},
                users: initialState.Cases.users
            },
            list: CASE.UNASSIGNED,
            disableItemList: initialState.Screens.All.disableItemList
        });
        expect(viewUnassigned.instance().state.usersRequested).toEqual(false);
    });

    it("Should call onRefresh method", () => {
        expect(viewUnassigned.instance().onRefresh()).toMatchSnapshot();
        expect(viewUnassigned.instance().onRefresh({nativeEvent: {text: "abc"}})).toMatchSnapshot();
        expect(viewUnassignedOffline.instance().onRefresh()).toMatchSnapshot();
    });

    it("Should call changeLayout method", () => {
        viewUnassigned.instance().changeLayout();
        expect(viewUnassigned.instance().state.middleHeight).toEqual(617);
    });

    it("Should call goToNewCases method", () => {
        expect(viewUnassigned.instance().goToNewCases()).toMatchSnapshot();
    });

    it("Should call refresh method of FlatList component", () => {
        const flatList = viewUnassigned.props("children").children[0];
        expect(flatList.props.renderItem({})).toMatchSnapshot();
        expect(flatList.props.onRefresh()).toMatchSnapshot();
    });

    it("Should call refresh method of FAB component", () => {
        const floatActionButton = viewUnassigned.props("children").children[1];
        expect(floatActionButton.props.onClickAction()).toMatchSnapshot();
    });
});
