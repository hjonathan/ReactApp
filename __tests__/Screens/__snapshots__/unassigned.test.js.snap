// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen >>> Unassigned Check static navigationOptions 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "padding": 16,
    }
  }
>
  <Icon
    allowFontScaling={false}
    name="bars"
    size={12}
    style={
      Object {
        "color": "#FFF",
        "fontSize": 20,
      }
    }
  />
</TouchableOpacity>
`;

exports[`Screen >>> Unassigned Render unassigned correctly 1`] = `
<Unassigned
  cases={
    Object {
      "unassigned": Array [
        Object {
          "prevUser": Object {
            "userId": "00001",
          },
        },
        Object {
          "prevUser": Object {
            "userId": "00002",
          },
        },
      ],
      "users": Array [
        Object {
          "userId": "00001",
          "userPhoto": "picture1",
        },
        Object {
          "userId": "00002",
          "userPhoto": "picture2",
        },
      ],
    }
  }
  isConnected={true}
  list="Unassigned"
  requestCasesUnassigned={[Function]}
  reset={[Function]}
  routeName={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  userInfoRequest={[Function]}
/>
`;

exports[`Screen >>> Unassigned Should call goToNewCases method 1`] = `undefined`;

exports[`Screen >>> Unassigned Should call onRefresh method 1`] = `undefined`;

exports[`Screen >>> Unassigned Should call onRefresh method 2`] = `undefined`;

exports[`Screen >>> Unassigned Should call onRefresh method 3`] = `undefined`;

exports[`Screen >>> Unassigned Should call refresh method of FAB component 1`] = `undefined`;

exports[`Screen >>> Unassigned Should call refresh method of FlatList component 1`] = `
<ItemList
  isConnected={true}
  parent={
    Object {
      "isConnected": true,
      "navigation": Object {
        "dispatch": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "routeName": "NewCases",
                "type": "Navigation/NAVIGATE",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "setParams": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "searchInput": false,
              },
            ],
            Array [
              Object {
                "searchInput": false,
              },
            ],
            Array [
              Object {
                "searchInput": false,
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "state": Object {
          "params": Object {
            "searchInput": null,
          },
          "routeName": "Draft",
        },
      },
      "requestCasesUnassigned": [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {
              "search": "",
            },
          ],
          Array [
            Object {
              "search": "abc",
            },
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "reset": [MockFunction],
      "routeName": [MockFunction] {
        "calls": Array [
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
          Array [
            "Draft",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "userInfoRequest": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "size": "1",
              "user": Object {
                "ids": Array [
                  "00001",
                  "00002",
                ],
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
/>
`;

exports[`Screen >>> Unassigned Should call refresh method of FlatList component 2`] = `undefined`;

exports[`Screen >>> Unassigned mapStateToProps 1`] = `
Object {
  "type": "cases/unassigned/request",
}
`;

exports[`Screen >>> Unassigned mapStateToProps 2`] = `
Object {
  "type": "cases/userInfo/request",
}
`;

exports[`Screen >>> Unassigned mapStateToProps 3`] = `
Object {
  "type": "cases/routeName",
}
`;

exports[`Screen >>> Unassigned mapStateToProps 4`] = `
Object {
  "type": "cases/reset",
}
`;
